CorsixTH/Lua/entities/humanoids/vip.lua

```diff
@@ -202,268 +290,240 @@ function Vip:announce()
-  end
-
-  -- if the player is about to kill a live patient for research, lower their rating dramatically
-  if room.room_info.id == "research" then
-    if room:getPatient() then
-      self.vip_rating = self.vip_rating - 80
+function Vip:setVIPRating()
+  -- first do room code for later
+  local count_rooms = 0
+  local sum_queue = 0
+  for _, room in pairs(self.world.rooms) do
+    if not room.crashed then
+      count_rooms = count_rooms + 1
+    end
+    if room.door.queue then
+      sum_queue = sum_queue + room.door.queue:size()
     end
   end
 
-  if room.staff_member then
-    if room.staff_member.profile.skill > 0.9 then
-      self.room_eval = self.room_eval + 3
-    end
-    if room.staff_member.attributes["fatigue"] then
-      if room.staff_member.attributes["fatigue"] < 0.4 then
-        self.room_eval = self.room_eval + 2
-      end
-    end
+--[[-- Group factor 1: Litter--]]
+  if (self.num_vomit_noninducing + self.num_vomit_inducing) <= 10 then
+    self.vip_rating = self.vip_rating - 1
+  else
+    self.vip_rating = self.vip_rating + 1
   end
 
-  -- evaluate the room we're currently looking at
-  for object, _ in pairs(room.objects) do
-    if object.object_type.id == "extinguisher" then
-      self.room_eval = self.room_eval + 1
-      break
-    elseif object.object_type.id == "plant" then
-      if object.days_left >= 10 then
-        self.room_eval = self.room_eval + 1
-      elseif object.days_left <= 3 then
-        self.room_eval = self.room_eval - 1
+--[[-- Group factor 2: Staff tiredness--]]
+  -- First get staff members
+  local count_staff = #self.hospital.staff
+  if count_staff > 1 then
+    -- Loop through staff tiredness, if any above verytired, break loop
+    for _, staff in ipairs(self.hospital.staff) do
+      if staff.attributes["fatigue"] ~= nil and staff.attributes["fatigue"] >= 0.7 then
+        self.vip_rating = self.vip_rating + 2
+        break
       end
-      break
     end
-
-    if object.strength then
-      if object.strength > (object.object_type.default_strength / 2) then
-        self.room_eval = self.room_eval + 1
-      else
-        self.room_eval = self.room_eval - 3
-      end
+    -- Average all staff tiredness. If above tiredness level award 1, else -1
+    local avg_tired = self.hospital:getAverageStaffAttribute("fatigue", 1)
+    if avg_tired >= 0.5 then
+      self.vip_rating = self.vip_rating + 1
+    else
+      self.vip_rating = self.vip_rating - 1
     end
+  else
+    -- Penalise where there is one or no staff
+    self.vip_rating = self.vip_rating + 4
   end
-end
 
-function Vip:evaluateEmergency(success)
-  -- Make sure that the VIP is still actually in the process of evaluation
-  if not self.going_home then
-    if success then
-      self.vip_rating = self.vip_rating + 10
+--[[-- Group factor 3: Patients--]]
+  -- first check we had patients this visit
+  local visitors_diff = self.hospital.num_visitors + self.enter_patients - self.enter_visitors
```

Comment:
Not clear why we care about visitor_diff , and I don't think this line gets the visitor diff. Don't we just care that there are patients in the hospital?

Reply:
The difficulty I noticed with visitors_diff is in its old definition it only counted who entered from when the VIP arrived -- they didn't care about who was already here (and thus usually the people who are cured during the visit). Therefore it would cause a bias on the other factors. 

The variable should be renamed to reflect that it no longer just pulls who entered over the visit. It counts who came + who was already here

Reply:
I don't think it does that right now. It counts who is currently there plus who came.

If you only care about the point in time the VIP evaluation happens you don't need to store any counts in the vip.

Reply:
The self.enter_patients measures who is in the hospital when the VIP arrives (done in world.lua). Then it adds on the number of visitors who came to the hospital during the visit. It doesn't check who is there at the time of eval.

Reply:
Talk on irc later.